src/main/c/NativeModule.cpp

```diff
@@ -0,0 +1,559 @@
+    }
+    return (rc == 0);
+  }
+};
+
+NativeModule::NativeModule(
+  const char* options,
+  const char* source,
+  const char* include
+) :
+  build_state_(kInit),
+  load_state_(kInit),
+  key_(hash_two_strings(options, source)),
+  is_global_(false),
+  dlopen_handle_(nullptr),
+  lib_name_(config.get_lib_name(key_)) {
+  // Master constructor - try to get module built in local file
+  config.ensure_module_dir_exists();
+  if (file_exists(lib_name_)) {
+    build_state_ = kPass;
+  } else {
+    // The file doesn't exist, let's build it
+    ModuleBuilder builder(options, source, include, key_);
+    build_state_ = (builder.try_to_build() ? kPass : kFail);
+  }
+}
+
+NativeModule::NativeModule(
+  bool is_global,
+  const char* key,
+  ssize_t binary_size,
+  const void* binary
+) :
+  build_state_(is_global ? kPass : kInit),
+  load_state_(is_global ? kPass : kInit),
+  key_(key),
+  is_global_(is_global),
+  dlopen_handle_(nullptr),
+  lib_name_(config.get_lib_name(key_)) {
+  // Worker constructor - try to get the binary written to local file
+  if (is_global_) return;
+  int rc = 0;
+  config.ensure_module_dir_exists();
+  if (!file_exists(lib_name_)) {
+    // We hold big_mutex so there is no race
+    int fd = open(lib_name_.c_str(), O_WRONLY|O_CREAT|O_TRUNC, 0666);
+    if (fd < 0) {
+      perror("open");
+      assert(0);
+    }
+    rc = write(fd, binary, binary_size);
+    assert(rc == binary_size);
+    ::close(fd);
+    ::chmod(lib_name_.c_str(), 0644);
+  }
+  build_state_ = kPass;
+  try_load_locked();
+}
+
+NativeModule::~NativeModule() {
+  if (!is_global_ && dlopen_handle_) {
+    dlclose(dlopen_handle_);
+  }
+}
+
+bool NativeModule::try_load_locked() {
+  if (load_state_ == kInit) {
+    if (is_global_) {
+      load_state_ = kPass;
+    } else {
+      auto handle = dlopen(lib_name_.c_str(), RTLD_GLOBAL|RTLD_NOW);
+      if (!handle) {
+        fprintf(stderr, "ERROR: dlopen failed: %s\n", dlerror());
+      }
+      load_state_ = (handle ? kPass : kFail);
+      if (handle) dlopen_handle_ = handle;
+    }
+  }
+  return (load_state_ == kPass);
+}
+
+std::vector<char> NativeModule::get_binary() {
+  std::lock_guard<std::mutex> mylock(big_mutex);
+  int fd = open(config.get_lib_name(key_).c_str(), O_RDONLY, 0666);
+  std::vector<char> vec;
+  if (fd < 0) {
+    return vec; // build failed, no lib, return empty
+  }
+  struct stat st;
+  int rc = fstat(fd, &st);
+  assert(rc == 0);
+  size_t file_size = st.st_size;
+  vec.resize(file_size);
+  rc = read(fd, &vec[0], file_size);
+  assert(rc == (int)file_size);
+  close(fd);
+  return vec;
+}
+
+static std::string to_qualified_name(
+  JNIEnv* env,
+  const std::string& key,
+  jstring nameJ,
+  int numArgs,
+  bool is_global,
+  bool is_longfunc
+) {
+  JString name(env, nameJ);
+  char argTypeCodes[32];
+  for (int j = 0; j < numArgs; ++j) argTypeCodes[j] = 'l';
+  argTypeCodes[numArgs] = 0;
+  char buf[512];
```

Comment:
This screams buffer overflow bug.  At least, use `snprintf` and the assert the output fit.